beps/0002-dynamic-frontend-plugins/README.md

```diff
@@ -154,12 +154,300 @@ Each plugin is required to provide manifest file (metadata) in predefined format
+- Installed node modules **can't be shared** to avoid dependency conflicts. Module federation is able to handle duplicate dependency versions.
+- Each plugin can be viewed as a `private` workspace package in a npm workspaces environment.
+
+```sh
+path/to/assets/plugin/dynamic-plugins/
+├── dynamic-backstage.plugin-foo
+│   ├── dist
+│   ├── node_modules # private_modules for plugins that need to be dynamized!
+│   ├── package.json
+│   ├── package-lock.json
+│   └── plugin-manifest.json
+└── dynamic-backstage.plugin-user-settings
+    ├── dist
+    ├── package.json
+    └── plugin-manifest.json
+```
+
+#### Boot time initialization
+
+**Dynamically build plugins**
+
+Plugins that already contain dynamic output. Including the [manifest](#plugin-manifest). No additional init required.
+
+**Static plugins**
+
+Have to be ["Dynamized"](#dynamize-plugin-during-backstage-boot) first. Plugin dependencies has to be installed. The new [Dynamic export CLI](#dynamic-export-cli) has to be called first.
+
+**Plugin registry collection**
+
+Once plugins are installed and fully dynamic, the asset server has to "discover" what plugins are loadable.
+
+This can be done by scanning the "assets" (dynamic-plugins in the dir example) for individual plugin manifests. Simple `glob` lookup for `path/dynamic-plugins/**/plugins-manifest.json` pattern should give required results.
+
+With this data the assets server should be able to generate a "plugin registry" with necessary metadata for the frontend. See the [Plugin Registry](#plugin-registry-1) for format details.
+
+The during the collection, the program should also add the `assetHost` prefix to the manifest location attribute described in the [webpack configuration](#sample-webpack-configuration) section.
+
+**Singleton dependencies verification (optional)**
+
+With [module sharing](#module-sharing) and its singleton option, it is critical that the singleton dependencies versions are matching.
+
+The singleton configuration from the shell application can be extracted as well as from all dynamic plugins (included in the generated [manifest](#plugin-manifest)). A semver check can be used to prevent a plugin from initializing if a version mismatch is found. A version check also has to be made between plugins. New plugin singleton dependencies must match both the shell and all previously loaded plugins.
+
+#### Runtime updates
+
+It is fully expected that plugins can be added or removed at the asset server runtime. The same tasks can be performed as during boot time initialization. Assigning a FS watcher (like [Chokidar](https://github.com/paulmillr/chokidar)) can be used to handle updates.
+
+#### Asset server API
+
+> NOTE Routes names are purely arbitrary at this point.
+
+The API can be very simplistic
+
+_GET /dynamic-plugins/plugin-registry_
+
+Returns generated [Plugin Registry](#plugin-registry-1).
+
+\*GET /dynamic-plugins/assets/\*\*
```

Comment:
We had similar discussions about bundles. I think "bundles" should be just an abstraction layer. It would be up to the plugin builders to define "bundles" or dependencies.

The question is, what is supposed to happen with a plugin if a dependency is missing? All plugins are loaded at boot time (in the frontend) technically all dependencies should be already present in the backend. I think the concept of bundles is just a cosmetic grouping for some plugin marketplace rather than something necessary for the API itself. At most, the API should only validate if a plugin dependency is also available.